File: driver/src/mailbox.rs
const MMIO_BASE: u32 = 0x3f000000;
const MAILBOX_BASE: u32 = MMIO_BASE + 0xb880;

const MAILBOX_READ: u32 = MAILBOX_BASE;
const MAILBOX_STATUS: u32 = MAILBOX_BASE + 0x18;
const MAILBOX_WRITE: u32 = MAILBOX_BASE + 0x20;

const MAILBOX_EMPTY: u32 = 0x40000000;
const MAILBOX_FULL: u32 = 0x80000000;

use core::ptr::{read_volatile, write_volatile};

#[no_mangle]
#[inline(never)]
pub fn mailbox_call(ch:u32, mailbox: *const u32) {
    // check if the mailbox is full
    unsafe {
        let mut status: u32;
        loop {
            status = read_volatile(MAILBOX_STATUS as *const u32);
            if (status & MAILBOX_FULL) == 0 {
                // mailbox is not full, we can write the message
                break;
            }
        }
        // Combine the message address (upper 28 bits) with channel number (lower 4 bits)
        let channel = ch;
        let message_address = mailbox as u32;
        let combined_address = (message_address & 0xFFFFFFF0) | (channel & 0xF);
        write_volatile(MAILBOX_WRITE as *mut u32, combined_address as u32);

        // check if the mailbox is empty
        loop {
            status = read_volatile(MAILBOX_STATUS as *const u32);
            if (status & MAILBOX_EMPTY) == 0 {
                // mailbox is not empty, we can read the message
                break;
            }
        }
        if read_volatile(MAILBOX_READ as *const u32) != (message_address as u32) {}
    }
}



const GET_BOARD_REVISION: u32 = 0x00010002;
const REQUEST_CODE: u32 = 0x00000000;
const REQUEST_SUCCEED: u32 = 0x80000000;
const REQUEST_FAILED: u32 = 0x80000001;
const TAG_REQUEST_CODE: u32 = 0x00000000;
const END_TAG: u32 = 0x00000000;

pub fn get_board_revisioin() -> u32 {
    let mut mailbox: [u32; 32] = [0; 32];
    let idx_offset: usize = (0x10 - mailbox.as_ptr() as u32 & 0xf) as usize;

    mailbox[0 + idx_offset] = 7 * 4;
    mailbox[1 + idx_offset] = REQUEST_CODE;
    mailbox[2 + idx_offset] = GET_BOARD_REVISION;
    mailbox[3 + idx_offset] = 4;
    mailbox[4 + idx_offset] = TAG_REQUEST_CODE;
    mailbox[5 + idx_offset] = 0;
    mailbox[6 + idx_offset] = END_TAG;

    mailbox_call(8,(mailbox[idx_offset..(idx_offset + 7)]).as_ptr());
    mailbox[5]
}

// Get ARM memory base address and size using mailbox
const GET_ARM_MEMORY: u32 = 0x00010005;
pub fn get_arm_memory() -> (u32, u32) {
    let mut mailbox: [u32; 32] = [0; 32];
    let idx_offset: usize = (0x10 - mailbox.as_ptr() as u32 & 0xf) as usize;

    mailbox[0 + idx_offset] = 8 * 4;
    mailbox[1 + idx_offset] = REQUEST_CODE;
    mailbox[2 + idx_offset] = GET_ARM_MEMORY;
    mailbox[3 + idx_offset] = 8;
    mailbox[4 + idx_offset] = TAG_REQUEST_CODE;
    mailbox[5 + idx_offset] = 0;
    mailbox[6 + idx_offset] = 0;
    mailbox[7 + idx_offset] = END_TAG;

    mailbox_call(8, mailbox[idx_offset..(idx_offset + 8)].as_ptr());
    (mailbox[5], mailbox[6])
}

File: lab3/src/mailbox.c
#include "../include/mailbox.h"
#include "../include/mini_uart.h"



/* mailbox message buffer */
volatile unsigned int  __attribute__((aligned(16))) mbox[8];

int mbox_call()
{
    /* 1. Combine the message address (upper 28 bits) with channel number (lower 4 bits) */
    unsigned int readchannel = (((unsigned int)((unsigned long)&mbox) & ~0xF) | (0x8 & 0xF));

    /* 2. Check if Mailbox 0 status register’s full flag is set.*/
    while (*MAILBOX_STATUS & MAILBOX_FULL);

    /* 3. If not, then you can write to Mailbox 1 Read/Write register.*/
    *MAILBOX_WRITE = readchannel;

    /* 4. Check if Mailbox 0 status register’s empty flag is set.*/
    while (1) {
        while (*MAILBOX_STATUS & MAILBOX_EMPTY);
        if (readchannel == *MAILBOX_READ) {
            return mbox[1] == MAILBOX_RESPONSE;
        }
    }
    return 0;
}

void get_board_revision()
{
    uart_send_string("In get_board_revision\r\n");
    mbox[0] = 7 * 4; // buffer size in bytes
    mbox[1] = REQUEST_CODE;
    // tags begin
    mbox[2] = GET_BOARD_REVISION; // tag identifier
    mbox[3] = 4;                  // maximum of request and response value buffer's length.
    mbox[4] = TAG_REQUEST_CODE;
    mbox[5] = 0; // value buffer
    // tags end
    mbox[6] = END_TAG;
    mbox_call(); // message passing procedure call, you should implement it following the 6 steps provided above.
    uart_send_string("Board revision: ");
    uart_send_string("0x");
    uart_hex(mbox[5]);
    uart_send_string("\r\n");
}

void get_base_address()
{
    mbox[0] = 8 * 4; // buffer size in bytes
    mbox[1] = REQUEST_CODE;
    // tags begin
    mbox[2] = ARM_MEMORY; // tag identifier
    mbox[3] = 8;          // maximum of request and response value buffer's length.
    mbox[4] = TAG_REQUEST_CODE;
    mbox[5] = 0; // value buffer
    mbox[6] = 0; // value buffer
    // tags end
    mbox[7] = END_TAG;
    mbox_call(); // message passing procedure call, you should implement it following the 6 steps provided above.
    uart_send_string("Arm base address: ");
    uart_send_string("0x");
    uart_hex(mbox[5]);
    uart_send_string("\r\n");
    uart_send_string("Arm memory size: ");
    uart_send_string("0x");
    uart_hex(mbox[6]);
    uart_send_string("\r\n");
}